### PR TITLE
documentation #631: added swagger documentation for 7 roles endpoints

### LINKED ISSUE #631: Add Swagger documentation to Role endpoints
Codestyle: https://gist.github.com/Hahhhayam/43197ad702006f475eb8d97fb8c08814
Example: #541
Needed endpoints (7): 
image
Mention: RoleController never documentated, so, you should add @ApiTags decorator above controller class

### PR DESCRIPTION
**Pull request by illiakroshka**
Get:
- /roles
- /roles/{roleId}
- /roles/{roleId}/grants

Post:
- /roles
- /roles/{roleId}/grants

Patch:
- /roles/{roleId}

closes #631

### CODE REVIEW

File: src/v2/api/controllers/RoleController.ts [hunk 1]
```diff
@@ -32,6 +47,17 @@ export class RoleController {
...
+  @ApiOkResponse({
+    type: RoleResponse,
+  })
+  @ApiForbiddenResponse({
+    description: `\n 
+                  NoPermissionException:
+                  You do not have permission to perform this action
+                  
+                  NoPermissionException:
+                  Your parent role weight is bigger than role weight`,
```
Comment (basarabst) [inline src/v2/api/controllers/RoleController.ts:59]: doesn't exist

File: src/v2/api/controllers/RoleController.ts [hunk 2]
```diff
@@ -32,6 +47,17 @@ export class RoleController {
   }
 
   @Access('roles.create')
+  @ApiOkResponse({
+    type: RoleResponse,
+  })
+  @ApiForbiddenResponse({
+    description: `\n 
+                  NoPermissionException:
+                  You do not have permission to perform this action
```
Comment (basarabst) [inline src/v2/api/controllers/RoleController.ts:56]: too mauch spaces
pattern: https://gist.github.com/Hahhhayam/43197ad702006f475eb8d97fb8c08814

File: src/v2/api/controllers/RoleController.ts [hunk 3]
```diff
@@ -32,6 +47,17 @@ export class RoleController {
   }
 
   @Access('roles.create')
```
Comment (basarabst) [inline src/v2/api/controllers/RoleController.ts:49]: add 
image

File: src/v2/api/controllers/RoleController.ts [hunk 4]
```diff
@@ -42,6 +68,19 @@ export class RoleController {
   }
 
   @Access('roles.$roleId.grants.create')
+  @ApiOkResponse({
+    type: GrantResponse,
+  })
+  @ApiBadRequestResponse({
+    description: `\n
+                  InvalidEntityIdException:  
+                  role with such id is not found`,
```
Comment (basarabst) [inline src/v2/api/controllers/RoleController.ts:77]: doesn't exist

File: src/v2/api/controllers/RoleController.ts [hunk 5]
```diff
@@ -42,6 +68,19 @@ export class RoleController {
...
+  })
+  @ApiForbiddenResponse({
+    description: `\n 
+                  NoPermissionException:
+                  You do not have permission to perform this action`,
```
Comment (basarabst) [inline src/v2/api/controllers/RoleController.ts:82]: remove spaces

File: src/v2/api/controllers/RoleController.ts [hunk 6]
```diff
@@ -50,6 +89,14 @@ export class RoleController {
     return this.roleService.createGrants(roleId, body.grants);
   }
 
+  @ApiOkResponse({
+    type: MappedGrantsResponse,
+  })
+  @ApiBadRequestResponse({
+    description: `\n
+                  InvalidEntityIdException:  
+                  role with such id is not found`,
```
Comment (basarabst) [inline src/v2/api/controllers/RoleController.ts:98]: doesn't exist

File: src/v2/api/controllers/RoleController.ts [hunk 7]
```diff
@@ -60,6 +107,18 @@ export class RoleController {
   }
 
   @Access('roles.$roleId.delete')
+  @ApiBearerAuth()
+  @ApiOkResponse()
+  @ApiBadRequestResponse({
+    description: `\n
+                  InvalidEntityIdException: 
+                  role with such id is not found`,
```
Comment (basarabst) [inline src/v2/api/controllers/RoleController.ts:115]: doesn't exist

File: src/v2/api/controllers/RoleController.ts [hunk 8]
```diff
@@ -60,6 +107,18 @@ export class RoleController {
   }
 
   @Access('roles.$roleId.delete')
```
Comment (basarabst) [inline src/v2/api/controllers/RoleController.ts:109]: add UnauthorizedException

File: src/v2/api/controllers/RoleController.ts [hunk 9]
```diff
@@ -60,6 +107,18 @@ export class RoleController {
...
+  @ApiOkResponse()
+  @ApiBadRequestResponse({
+    description: `\n
+                  InvalidEntityIdException: 
+                  role with such id is not found`,
+  })
+  @ApiForbiddenResponse({
+    description: `\n
+                  NoPermissionException:
+                  You do not have permission to perform this action`,
```
Comment (basarabst) [inline src/v2/api/controllers/RoleController.ts:120]: spaces

File: src/v2/api/controllers/RoleController.ts [hunk 10]
```diff
@@ -69,6 +128,19 @@ export class RoleController {
   }
 
   @Access('roles.$roleId.update')
+  @ApiOkResponse({
+    type: UpdatedRoleResponse,
+  })
+  @ApiBadRequestResponse({
+    description: `\n
+                  InvalidEntityIdException:  
+                  role with such id is not found`,
```
Comment (basarabst) [inline src/v2/api/controllers/RoleController.ts:137]: doesn't exist

File: src/v2/api/controllers/RoleController.ts [hunk 11]
```diff
@@ -69,6 +128,19 @@ export class RoleController {
...
+  })
+  @ApiForbiddenResponse({
+    description: `\n
+                  NoPermissionException:
+                  You do not have permission to perform this action`,
+  })
```
Comment (basarabst) [inline src/v2/api/controllers/RoleController.ts:143]: remove spaces

File: src/v2/api/responses/GrantResponse.ts [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+export class MappedGrant {
+  @ApiProperty()
+  id: string;
+
+  @ApiProperty()
+  set: boolean;
+
+  @ApiProperty()
+  permission: string;
```
Comment (basarabst) [inline src/v2/api/responses/GrantResponse.ts:12]: linter

File: src/v2/api/responses/GrantResponse.ts [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+export class MappedGrantsResponse {
+  @ApiProperty({ type: [MappedGrant] })
+  grants: MappedGrant[];
+}
+
+export class GrantResponse {
+  @ApiProperty({ type: [Grant] })
+  grants: Grant[];
```
Comment (basarabst) [inline src/v2/api/responses/GrantResponse.ts:30]: linter

**Review by basarabst — CHANGES_REQUESTED** [review]

File: src/v2/api/dtos/CreateRoleWithGrantsDTO.ts [hunk 14]
```diff
@@ -1,14 +1,17 @@
+import { ApiPropertyOptional } from '@nestjs/swagger';
 import { IsOptional, IsUUID, ValidateNested } from 'class-validator';
 import { Type } from 'class-transformer';
 import { CreateRoleDTO } from './CreateRoleDTO';
 import { CreateGrantDTO } from './CreateGrantsDTO';
 
 export class CreateRoleWithGrantsDTO extends CreateRoleDTO {
+  @ApiPropertyOptional()
```
Comment (basarabst) [inline src/v2/api/dtos/CreateRoleWithGrantsDTO.ts:8]:
```
@ApiPropertyOptional({
    type: [CreateGrantDTO],
  })
  ```

File: src/v2/api/dtos/CreateRoleDTO.ts [hunk 15]
```diff
@@ -1,12 +1,15 @@
+import { ApiProperty } from '@nestjs/swagger';
 import { RoleName } from '@prisma/client';
 import { IsEnum, IsNotEmpty, IsNumber } from 'class-validator';
 import { validationOptionsMsg } from '../../utils/GLOBALS';
 
 export class CreateRoleDTO {
+  @ApiProperty()
```
Comment (basarabst) [inline src/v2/api/dtos/CreateRoleDTO.ts:7]: add
```
{ enum: RoleName }
```

File: src/v2/api/responses/GrantResponse.ts [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+  @ApiProperty()
+    permission: string;
+}
+
+export class Grant extends MappedGrant {
+  @ApiProperty()
+    roleId: string;
+
+  @ApiProperty()
```
Comment (basarabst) [inline src/v2/api/responses/GrantResponse.ts:19]: ``` type: Role ```

File: src/v2/api/controllers/RoleController.ts [hunk 17]
```diff
@@ -42,6 +71,20 @@ export class RoleController {
   }
 
   @Access('roles.$roleId.grants.create')
+  @ApiBearerAuth()
+  @ApiOkResponse({
+    type: GrantResponse,
+  })
```
Comment (basarabst) [inline src/v2/api/controllers/RoleController.ts:77]: wrong response, see it in Postman

File: src/v2/api/controllers/RoleController.ts [hunk 18]
```diff
@@ -60,6 +106,18 @@ export class RoleController {
   }
 
   @Access('roles.$roleId.delete')
```
Comment (basarabst) [inline src/v2/api/controllers/RoleController.ts:108]: where is response?

File: src/v2/api/responses/RoleResponse.ts [hunk 19]
```diff
@@ -0,0 +1,35 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { MappedGrant } from './GrantResponse';
+import { RoleName } from '@prisma/client';
+
+export class UpdatedRoleResponse {
+  @ApiProperty()
+    id: string;
+
+  @ApiProperty()
```
Comment (basarabst) [inline src/v2/api/responses/RoleResponse.ts:9]: add
```
{ type: RoleName }
```

**Review by basarabst — CHANGES_REQUESTED** [review]

**Review by basarabst — CHANGES_REQUESTED** [review]
Commit message is wrong, see example:
image

**Review by basarabst — CHANGES_REQUESTED** [review]
you have to squash all the commits to one, after that do a force push

File: src/v2/api/dtos/UpdateRoleDTO.ts [hunk 20]
```diff
@@ -1,10 +1,16 @@
+import { ApiPropertyOptional } from '@nestjs/swagger';
 import { RoleName } from '@prisma/client';
 import { IsOptional } from 'class-validator';
 
+
```
Comment (basarabst) [inline src/v2/api/dtos/UpdateRoleDTO.ts:5]: redundant indentation

**Review by basarabst — CHANGES_REQUESTED** [review]

**Review by basarabst — APPROVED** [review]

File: src/v2/api/dtos/CreateGrantsDTO.ts [hunk 21]
```diff
@@ -1,18 +1,24 @@
...
-  @ValidateNested({ each: true })
-    grants: CreateGrantDTO[];
-}
-
 export class CreateGrantDTO {
+  @ApiProperty()
   @IsNotEmpty(validationOptionsMsg('Permission can not be empty'))
     permission: string;
 
+  @ApiPropertyOptional()
```
Comment (ArtemLyah) [inline src/v2/api/dtos/CreateGrantsDTO.ts:11]: Add default true like there
image

File: src/v2/api/controllers/RoleController.ts [hunk 22]
```diff
@@ -33,6 +47,25 @@ export class RoleController {
...
+    description: `\n
+    UnauthorizedException:
+      Unauthorized`,
+  })
+  @ApiForbiddenResponse({
+    description: `\n 
+    NoPermissionException:
+      You do not have permission to perform this action`,
+  })
+  @ApiBadRequestResponse({
```
Comment (ArtemLyah) [inline src/v2/api/controllers/RoleController.ts:64]: the CreateRoleWithGrantsDTO uses CreateRoleDTO and CreateGrantDTO so add the following fields
image
image
like this
image

File: src/v2/api/controllers/RoleController.ts [hunk 23]
```diff
@@ -61,6 +111,21 @@ export class RoleController {
   }
 
   @Access(PERMISSION.ROLES_$ROLEID_DELETE)
+  @Access('roles.$roleId.delete')
```
Comment (ArtemLyah) [inline src/v2/api/controllers/RoleController.ts:114]: delete the Access

File: src/v2/api/dtos/UpdateRoleDTO.ts [hunk 24]
```diff
@@ -1,10 +1,15 @@
...
 import { IsOptional } from 'class-validator';
 
 export class UpdateRoleDTO {
+  @ApiPropertyOptional({
+    enum: RoleName,
+  })
   @IsOptional()
     name?: RoleName;
 
+  @ApiPropertyOptional()
```
Comment (ArtemLyah) [inline src/v2/api/dtos/UpdateRoleDTO.ts:12]: Add IsNumber decorator and a validation option message
don't forget to add validation message to "update" endpoint to InvalidBodyException

**Pull request merged**